### PR TITLE
Add option to set responseTimeout and handshakeTimeout

### PR DESCRIPTION
**Pull request by banterCZ**
Add option to set `responseTimeout` and `handshakeTimeout` at `RestClientConfiguration`.

### CODE REVIEW

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java [hunk 1]
```diff
@@ -496,4 +503,39 @@ public void setDefaultHttpHeaders(HttpHeaders headers) {
         this.defaultHttpHeaders = headers;
     }
 
+    /**
+     * Get the SSL handshake timeout.
+     *
+     * @return timeout duration
+     */
+    public Optional<Duration> getHandshakeTimeout() {
```
Comment (romanstrobl) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java:511]: Why do we use `Optional` for this parameter, but we do not use it for other parameters?

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java [hunk 2]
```diff
@@ -110,13 +116,24 @@ private void initializeWebClient() throws RestClientException {
         final SslContext sslContext = SslUtils.prepareSslContext(config);
         HttpClient httpClient = HttpClient.create();
         if (sslContext != null) {
-            httpClient = httpClient.secure(sslContextSpec -> sslContextSpec.sslContext(sslContext));
+            httpClient = httpClient.secure(sslContextSpec -> {
+                final SslProvider.Builder sslProviderBuilder = sslContextSpec.sslContext(sslContext);
+                config.getHandshakeTimeout().ifPresent(it -> {
+                    logger.debug("Setting handshake timeout {}", it);
+                    sslProviderBuilder.handshakeTimeout(it);
```
Comment (romanstrobl) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:123]: Have we tested this configuration? I am a bit concerned that we are using a builder and the `build()` method is not called. Perhaps it works, but I would just like to doublecheck this.

**Review by romanstrobl — APPROVED** [review]
I approve the pull request, however I have two minor concerns for investigation.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java (hunk 1 above)
Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java:511]: I do not insist, it was just a suggestion how to do it. But yes, we ca be consistent.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java (hunk 2 above)
Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:123]: I will add a test, if you are ok with this suggestion.

Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:123]: Build method is called in
```java
	public final HttpClient secure(Consumer<? super SslProvider.SslContextSpec> sslProviderBuilder) {
		Objects.requireNonNull(sslProviderBuilder, "sslProviderBuilder");
		SslProvider.SslContextSpec builder = SslProvider.builder();
		sslProviderBuilder.accept(builder);
		SslProvider sslProvider = HttpClientSecure.sslProvider(((SslProvider.Builder) builder).build());
		if (sslProvider.equals(configuration().sslProvider)) {
			return this;
		}
		HttpClient dup = duplicate();
		dup.configuration().sslProvider = sslProvider;
		return dup;
	}
```

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java [hunk 3]
```diff
@@ -110,13 +116,24 @@ private void initializeWebClient() throws RestClientException {
...
         }
         if (config.getConnectionTimeout() != null) {
             httpClient = httpClient.option(
                     ChannelOption.CONNECT_TIMEOUT_MILLIS,
                     config.getConnectionTimeout());
         }
+        final Duration responseTimeout = config.getResponseTimeout();
+        if (responseTimeout != null) {
+            logger.debug("Setting response timeout {}", responseTimeout);
+            httpClient.responseTimeout(responseTimeout);
```
Comment (banterCZ) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:135]: Missing assignment, has to be fixed.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java (hunk 1 above)
Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java]: `Optional` removed.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java (hunk 2 above)
Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java]: Test added.

File: rest-client-base/src/test/java/com/wultra/core/rest/client/base/DefaultRestClientTest.java [hunk 4]
```diff
@@ -498,4 +502,26 @@ public void testDefaultHttpHeaders() throws RestClientException {
         assertEquals(headerVaue, responseEntity.getHeaders().getFirst(headerName));
     }
 
+    @Test
+    public void testConfiguration() throws RestClientException {
+        final RestClientConfiguration config = prepareConfiguration();
+        final DefaultRestClient restClient = new DefaultRestClient(config);
+
+        final Object handshakeTimeout = getField(restClient, "webClient.builder.connector.httpClient.config.sslProvider.handshakeTimeoutMillis");
```
Comment (banterCZ) [inline rest-client-base/src/test/java/com/wultra/core/rest/client/base/DefaultRestClientTest.java:510]: Quite fragile, but now, I do not see better option how to test.

File: rest-client-base/src/test/java/com/wultra/core/rest/client/base/DefaultRestClientTest.java [hunk 5]
```diff
@@ -498,4 +502,26 @@ public void testDefaultHttpHeaders() throws RestClientException {
         assertEquals(headerVaue, responseEntity.getHeaders().getFirst(headerName));
     }
 
+    @Test
+    public void testConfiguration() throws RestClientException {
```
Comment (banterCZ) [inline rest-client-base/src/test/java/com/wultra/core/rest/client/base/DefaultRestClientTest.java:506]: `public` is not needed in JUnit 5. Keeping it consistent so far. I will improve it in a separate PR.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java [hunk 6]
```diff
@@ -47,9 +50,11 @@ public RestClientConfiguration() {
 
     // HTTP connection timeout
     private Integer connectionTimeout = 5000;
+    private Duration responseTimeout;
 
     // TLS certificate settings
     private boolean acceptInvalidSslCertificate = false;
+    private Duration handshakeTimeout;
```
Comment (saalistaja) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java:57]: Shouldn't we document the default value here?

`SslProvider:335`
```static final long DEFAULT_SSL_HANDSHAKE_TIMEOUT = Long.parseLong(System.getProperty("reactor.netty.tcp.sslHandshakeTimeout", "10000"));```

**Review by saalistaja — APPROVED** [review]
Great! Thanks for those two more tuning attributes.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java (hunk 6 above)
Comment (banterCZ) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java:57]: Default value is documented at setter.

**Pull request merged**